### PR TITLE
feat: Add support for `pre-commit/pre-commit-hooks` in Docker image

### LINKED ISSUE #359: [Other pre-commit hooks in pre-commit-terraform Docker image] github action error during installation
Hi People, 

# Configurations 

I have this configurations

```yaml
repos:
  ## general
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.2.0
    hooks:
      # Common errors
      - id: end-of-file-fixer
        exclude_types: [sql]
      - id: trailing-whitespace
        args: [--markdown-linebreak-ext=md]
        exclude_types: [sql]
      - id: check-yaml
      - id: check-executables-have-shebangs
      # Cross platform
      - id: check-case-conflict
      - id: mixed-line-ending
        args: [--fix=lf]
        exclude_types: [sql]
      # Security
      - id: detect-aws-credentials
        args: ['--allow-missing-credentials']
      - id: detect-private-key
  ## terraform
  - repo: https://github.com/antonbabenko/pre-commit-terraform
    rev: v1.64.1
    hooks:
      - id: terraform_fmt
      - id: terraform_docs
        args:
          - --hook-config=--path-to-file=README.md        # Valid UNIX path. I.e. ../TFDOC.md or docs/README.md etc.
          - --hook-config=--add-to-existing-file=true     # Boolean. true or false
          - --hook-config=--create-file-if-not-exist=true # Boolean. true or false
          - --args=--hide providers
      - id: terraform_tfsec
      - id: terraform_validate
```

Pipeline

```yaml
name: Static Analysis

on:
  push:

jobs:
  static_analysis:
    runs-on: ubuntu-20.04

    steps:
      - uses: actions/checkout@v2

      - name: init_src_policy
        env:
          FOLDERS: policy
        run: |
          TAG=$(cat .terraform-version)
          for f in $FOLDERS; do
            cd src/$f
            sed -i -e 's/  backend "azurerm" {}//g' 99_main.tf # use local backend
            docker run -v $(pwd):/tmp -w /tmp hashicorp/terraform:$TAG init
          done

      - name: run_test
        run: |
          TAG=v1.64.1
          docker run -v $(pwd):/lint -w /lint ghcr.io/antonbabenko/pre-commit-terraform:$TAG run -a

```

# Problem

Running this action, i find this error

```sh
Run TAG=v1.64.1
Unable to find image 'ghcr.io/antonbabenko/pre-commit-terraform:v1.64.1' locally
v1.64.1: Pulling from antonbabenko/pre-commit-terraform
[5](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:5)9bf1c3509f3: Already exists
07a400e93df3: Pulling fs layer
d8c[6](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:6)81bc2183: Pulling fs layer
c[7](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:7)f1ca1171d9: Pulling fs layer
dc9256fcea2e: Pulling fs layer
[8](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:8)2b2235ac423: Pulling fs layer
31ced[9](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:9)5c34e7: Pulling fs layer
5a3e01799d8c: Pulling fs layer
123b07747a77: Pulling fs layer
dc9256fcea2e: Waiting
82b2235ac423: Waiting
31ced95c34e7: Waiting
5a3e01799d8c: Waiting
123b07747a77: Waiting
843dab449416: Pulling fs layer
843dab449416: Waiting
c7f1ca1171d9: Verifying Checksum
c7f1ca1171d9: Download complete
07a400e93df3: Verifying Checksum
07a400e93df3: Download complete
d8c681bc2183: Verifying Checksum
d8c681bc2183: Download complete
dc9256fcea2e: Verifying Checksum
dc9256fcea2e: Download complete
82b2235ac423: Verifying Checksum
82b2235ac423: Download complete
07a400e93df3: Pull complete
123b07747a77: Verifying Checksum
123b07747a77: Download complete
5a3e01799d8c: Verifying Checksum
5a3e01799d8c: Download complete
31ced95c34e7: Verifying Checksum
31ced95c34e7: Download complete
843dab449416: Verifying Checksum
843dab449416: Download complete
d8c681bc2183: Pull complete
c7f1ca1171d9: Pull complete
dc9256fcea2e: Pull complete
82b2235ac423: Pull complete
31ced95c34e7: Pull complete
5a3e01799d8c: Pull complete
123b07747a77: Pull complete
843dab449416: Pull complete
Digest: sha256:f1e4797d02a89ab274e984ff26154d799c1209a642cc6ab362973790b2b6182a
Status: Downloaded newer image for ghcr.io/antonbabenko/pre-commit-terraform:v1.64.1
[INFO] Initializing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Initializing environment for https://github.com/antonbabenko/pre-commit-terraform.
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: CalledProcessError: command: ('/root/.cache/pre-commit/repoibac5ccm/py_env-python3/bin/python', '-mpip', 'install', '.')
return code: 1
expected return code: 0
stdout:
    Processing /root/.cache/pre-commit/repoibac5ccm
      Preparing metadata (setup.py): started
      Preparing metadata (setup.py): finished with status 'done'
    Collecting ruamel.yaml>=0.15
      Downloading ruamel.yaml-0.17.21-py3-none-any.whl ([10](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:10)9 kB)
         ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 109.5/109.5 KB 8.2 MB/s eta 0:00:00
    Collecting toml
      Downloading toml-0.10.2-py2.py3-none-any.whl ([16](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:16) kB)
    Collecting ruamel.yaml.clib>=0.2.6
      Downloading ruamel.yaml.clib-0.2.6.tar.gz (180 kB)
         ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 180.7/180.7 KB 23.0 MB/s eta 0:00:00
      Preparing metadata (setup.py): started
      Preparing metadata (setup.py): finished with status 'error'
    
stderr:
      error: subprocess-exited-with-error
      
      × python setup.py egg_info did not run successfully.
      │ exit code: 1
      ╰─> [3 lines of output]
          sys.argv ['/tmp/pip-install-htjtaeow/ruamel-yaml-clib_c5e3ea74f3e94e83b1944fd834e[17](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:17)1[72](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true#step:4:72)/setup.py', 'egg_info', '--egg-base', '/tmp/pip-pip-egg-info-s8ovu009']
          test compiling /tmp/tmp_ruamel_at0lb4mm/test_ruamel_yaml.c -> test_ruamel_yaml compile error: /tmp/tmp_ruamel_at0lb4mm/test_ruamel_yaml.c
          Exception: command 'gcc' failed: No such file or directory
          [end of output]
      
      note: This error originates from a subprocess, and is likely not a problem with pip.
    error: metadata-generation-failed
    
    × Encountered error while generating package metadata.
    ╰─> See above for output.
    
    note: This is an issue with the package mentioned above, not pip.
    hint: See above for details.
    
Check the log at /root/.cache/pre-commit/pre-commit.log
Error: Process completed with exit code 3.

```
see: [pipeline log](https://github.com/pagopa/devops-azure-governance/runs/5991589876?check_suite_focus=true)

# This configuration works

```yaml
repos:
  ## terraform
  - repo: https://github.com/antonbabenko/pre-commit-terraform
    rev: v1.64.1
    hooks:
      - id: terraform_fmt
      - id: terraform_docs
        args:
          - --hook-config=--path-to-file=README.md        # Valid UNIX path. I.e. ../TFDOC.md or docs/README.md etc.
          - --hook-config=--add-to-existing-file=true     # Boolean. true or false
          - --hook-config=--create-file-if-not-exist=true # Boolean. true or false
          - --args=--hide providers
      - id: terraform_tfsec
      - id: terraform_validate

```
see <https://github.com/pagopa/devops-azure-governance/runs/5991612215?check_suite_focus=true>

# Help

Please can you help me to understand what is the problem, because in local works as expected.

ps: in the logs there are some info about the fact that gcc is not found, but i make some test and gcc is in place.

Have a nice day

### PR DESCRIPTION
**Pull request by Jno21**
Put an `x` into the box if that apply:

- [ ] This PR introduces breaking change.
- [x] This PR fixes a bug.
- [x] This PR adds new functionality.
- [ ] This PR enhances existing functionality.

### Description of your changes

Fixes #359 :
Add `alpine-sdk` package so we are able to run third party checks. This is mostly to be able to run directly this docker image with checks coming from `https://github.com/pre-commit/pre-commit-hooks` since they are widely use in combination with `pre-commit-terraform`.

This is to avoid all the users of the community using pre-commit-terraform to build their own image on top of the one existing.

FIY: This was working before the  v1.63.0. So I guess it is a change coming directly from the image of Python.

-->

### How can we test changes

1. Create a `.pre-commit-config.yaml` file containing:
```yaml
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.2.0
    hooks:
      # Common errors
      - id: end-of-file-fixer
      - id: trailing-whitespace
        args: [--markdown-linebreak-ext=md]
      - id: check-yaml

      # Security
      - id: detect-aws-credentials
        args: ["--allow-missing-credentials"]
      - id: detect-private-key

  - repo: https://github.com/antonbabenko/pre-commit-terraform
    rev: v1.68.1
    hooks:
      - id: terraform_fmt
      - id: terraform_docs
      - id: terraform_tfsec
      - id: terraform_tflint
      - id: terraform_validate
```
2. Build the docker image using my branch `docker build -t pre-commit-terraform --build-arg INSTALL_ALL=true .`
3. Run it `docker run -v $(pwd):/lint -it -w /lint pre-commit-terraform run -a`

All the tests should run.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -161,7 +161,8 @@ RUN apk add --no-cache \
     # pre-commit deps
     git=~2 \
     # All hooks deps
-    bash=~5
+    bash=~5 \
+    alpine-sdk
```
Comment (yermulnik) [inline Dockerfile:165]: Should the comment be added to this line to align with how other packages in this block are described?

**Review by yermulnik — APPROVED** [review]

File: Dockerfile (hunk 1 above)
Comment (Jno21) [reply Dockerfile:165]: I am not sure what you mean, do you mean should we add another comment to differentiate the hooks deps.
```suggestion
    # Third party hooks deps
    alpine-sdk
```

Like this ?

**Review by MaxymVlasov — COMMENTED** [review]
1. `alpine-sdk` is a too big package. pre-commit "default" hooks need https://wiki.alpinelinux.org/wiki/GCC
2. Depending on what you use, you need different hooks and deps. I oppose adding half of the internet to the Docker image because some folks besides TF can use Golang, Python, Bash, Ruby, YAML, etc. in their IaC.

So, first of all, we need to understand what we'd like to support and how, because what I see in this PR is not really is:
> Add alpine-sdk as package to be able to run third party hooks

**Comment by Jno21** [comment]
> 1. `alpine-sdk` is a too big package. pre-commit "default" hooks need https://wiki.alpinelinux.org/wiki/GCC
> 2. Depending on what you use, you need different hooks and deps. I oppose adding half of the internet to the Docker image because some folks besides TF can use Golang, Python, Bash, Ruby, YAML, etc. in their IaC.
> 
> So, first of all, we need to understand what we'd like to support and how, because what I see in this PR is not really is:
> 
> > Add alpine-sdk as package to be able to run third party hooks

1. You are right, `build-base` is enough to make it works. I am can update my PR with only this if you want.
2. I totally agree with this point, I knew this can will create a debate, like I said in the issue. My point of view on the idea is not to supports everything since it would be unmaintainable.
I think having at least support for hooks coming from `https://github.com/pre-commit/pre-commit-hooks` has some sense. They are belonging to the `pre-commit`, and they are widely use and can be useful with basic terraform projects.

It was working before with the image v1.62.3, even if it wasn't something that was specifically wanted for this image.

The question that I am asking then is could we support at least the hooks from `https://github.com/pre-commit/pre-commit-hooks`. That would be a huge gain for the community, since we we won't all need to create our own image on top.

I would understand any answer as I see the pros and cons adding this. We can discuss this further if you would like to.

File: Dockerfile (hunk 1 above)
Comment (yermulnik) [reply Dockerfile:165]: Yep, that's what I mean. Thanks.

**Comment by MaxymVlasov** [comment]
I suggest installing as minimum as possible deps.
Add all existing hooks in `pre-commit-hooks` to your test `.pre-commit-config.yaml` and try running them with `gcc` and `musl-dev or libc-dev` as suggested in the [Alpine GCC article](https://wiki.alpinelinux.org/wiki/GCC).

Let's find min req deps and add only them. With major version pining.

**Comment by Jno21** [comment]
Hi,

Thank you for your feedbacks. I update the PR so we only install `musl-dev` and `gcc`. I tested it with all the hooks and it seems to be working.

Tell me if you want me to change the title of the PR and rebase to only have one commit.

Thank you !

**Review by MaxymVlasov — APPROVED** [review]

**Review by antonbabenko — APPROVED** [review]

**Comment by antonbabenko** [comment]
This PR is included in version 1.70.0 :tada:

**Pull request merged**